ps2-decompile/attach.c

```diff
@@ -3,188 +3,170 @@
+    }
+    if (hdesc->bLength != 9 ||
+        hdesc->bDescriptorType != 0x21 ||
+        LOBYTE(hdesc->bcdUSB) != 0x10 ||
+        HIBYTE(hdesc->bcdUSB) != 1 ||
+        hdesc->bDeviceClass ||
+        hdesc->bDeviceSubClass != 1 ||
+        hdesc->bDeviceProtocol != 0x22 ||
+        hdesc->bMaxPacketSize0 != 0x52 ||
+        LOBYTE(hdesc->idVendor)) {
+        g->descriptorChecksPassed = 0;
+        printf("hdesc check failed\n");
+    }
+    if (edesc->bEndpointAddress != 0x81 ||
+        edesc->bmAttributes != 3 ||
+        edesc->wMaxPacketSizeLB != 8 ||
+        edesc->wMaxPacketSizeHB ||
+        LOBYTE(edesc->bInterval) != 4) {
+        g->descriptorChecksPassed = 0;
+        printf("edesc check failed\n");
+    }
+
+    if ((getDeviceLocationRet = sceUsbdGetDeviceLocation(
+             devId, deviceLocationPath))) {
+        printf("usbmouse%d: sceUsbdGetDeviceLocation -> 0x%x\n", g->port1,
+               getDeviceLocationRet);
+        unit_free(g);
+        return -1;
+    }
+
+    printf("usbmouse%d: attached (port=", g->port1);
+
+    unrolled(0xFFFF8048, deviceLocationPath[0], v5, gp0);
+    unrolled(0xFFFF8044, deviceLocationPath[1], v5, gp0);
+    unrolled(0xFFFF8044, deviceLocationPath[2], v5, gp0);
+    unrolled(0xFFFF8044, deviceLocationPath[3], v5, gp0);
+    unrolled(0xFFFF8044, deviceLocationPath[4], v5, gp0);
+    unrolled(0xFFFF8044, deviceLocationPath[5], v5, gp0);
+    unrolled(0xFFFF8044, deviceLocationPath[6], v5, gp0);
+
+    printf((const char *)(gp0 + 0xFFFF804C));
+    printf((const char *)(gp2 + 0xFFFF8020));
+    printf((const char *)(gp1 + 0xFFFF8050),deviceLocationPath[0]);
+    printf((const char *)(gp1 + 0xFFFF8050),deviceLocationPath[1]);
+    printf((const char *)(gp1 + 0xFFFF8050),deviceLocationPath[2]);
+    printf((const char *)(gp1 + 0xFFFF8050),deviceLocationPath[3]);
+    printf((const char *)(gp1 + 0xFFFF8050),deviceLocationPath[4]);
+    printf((const char *)(gp1 + 0xFFFF8050),deviceLocationPath[5]);
+    printf((const char *)(gp1 + 0xFFFF8050),deviceLocationPath[6]);
+    printf((const char *)(gp1 + 0xFFFF8020));
+
+    switch (deviceLocationPath[0]) {
+    case 1: {
+        g->port |= 2u;
+        printf("attached port = 1\n");
+        break;
+    }
+    case 2: {
+        g->port |= 1u;
+        printf("attached port = 2\n");
+        break;
+    }
+    }
+
+    printf("c_pipe = %d\n", g->configEndpoint);
+    deviceRequest.requesttype = 0; // USB_DIR_OUT
+    deviceRequest.request = 9; // USB_REQ_SET_CONFIGURATION
+    deviceRequest.value = cdesc->bConfigurationValue;
+    deviceRequest.index = 0;
+    deviceRequest.length = 0;
+
+    if ((transferPipeRet = sceUsbdTransferPipe(g->configEndpoint, 0, 0,
+                                               &deviceRequest, set_config_done, g))) {
+        printf("usbmouse%d: sceUsbdSetConfiguration -> 0x%x\n",
+               g->port1, transferPipeRet);
+        unit_free(g);
+        return -1;
+    }
+
+    dump_desc(devId);
+    unit_free(g);
```

Comment:
This should not be here. The `g` value needs to exist throughout the entire driver's lifespan. Only `usbmouse_detach` should free its memory.